src/rgw/driver/rados/topic.cc

```diff
@@ -0,0 +1,383 @@
+  r = rgw_rados_operate(dpp, ref.ioctx, ref.obj.oid, &op, nullptr, y);
+  if (r == -ENOENT) {
+    return 0;
+  }
+  if (r < 0) {
+    return r;
+  }
+  if (rval < 0) {
+    return rval;
+  }
+
+  if (more && !keys.empty()) {
+    next_marker = *keys.rbegin();
+  } else {
+    next_marker.clear();
+  }
+  bucket_keys.merge(std::move(keys));
+
+  return 0;
+}
+
+
+class MetadataObject : public RGWMetadataObject {
+  rgw_pubsub_topic info;
+public:
+  MetadataObject(const rgw_pubsub_topic& info, const obj_version& v, real_time m)
+    : RGWMetadataObject(v, m), info(info) {}
+
+  void dump(Formatter *f) const override {
+    info.dump(f);
+  }
+
+  rgw_pubsub_topic& get_topic_info() {
+    return info;
+  }
+};
+
+class MetadataLister : public RGWMetadataLister {
+ public:
+  using RGWMetadataLister::RGWMetadataLister;
+
+  virtual void filter_transform(std::vector<std::string>& oids,
+                                std::list<std::string>& keys) {
+    // remove the oid prefix from keys
+    constexpr auto trim = [] (const std::string& oid) {
+      return oid.substr(oid_prefix.size());
+    };
+    std::transform(oids.begin(), oids.end(),
+                   std::back_inserter(keys),
+                   trim);
+  }
+};
+
+class MetadataHandler : public RGWMetadataHandler {
+  RGWSI_SysObj& sysobj;
+  RGWSI_SysObj_Cache& cache_svc;
+  RGWSI_MDLog& mdlog;
+  const RGWZoneParams& zone;
+  RGWChainedCacheImpl<cache_entry>& cache;
+ public:
+  MetadataHandler(RGWSI_SysObj& sysobj, RGWSI_SysObj_Cache& cache_svc,
+                  RGWSI_MDLog& mdlog, const RGWZoneParams& zone,
+                  RGWChainedCacheImpl<cache_entry>& cache)
+    : sysobj(sysobj), cache_svc(cache_svc), mdlog(mdlog),
+      zone(zone), cache(cache)
+  {}
+
+  std::string get_type() final { return "topic";  }
+
+  RGWMetadataObject* get_meta_obj(JSONObj *jo,
+                                  const obj_version& objv,
+                                  const ceph::real_time& mtime) override
+  {
+    rgw_pubsub_topic info;
+
+    try {
+      info.decode_json(jo);
+    } catch (JSONDecoder:: err& e) {
+      return nullptr;
+    }
+
+    return new MetadataObject(info, objv, mtime);
+  }
+
+  int get(std::string& entry, RGWMetadataObject** obj,
+          optional_yield y, const DoutPrefixProvider* dpp) override
+  {
+    cache_entry e;
+    int ret = read(dpp, y, sysobj, cache_svc, zone, entry,
+                   e.info, cache, &e.mtime, &e.objv);
+    if (ret < 0) {
+      return ret;
+    }
+
+    *obj = new MetadataObject(e.info, e.objv.read_version, e.mtime);
+    return 0;
+  }
+
+  int put(std::string& entry, RGWMetadataObject* obj,
+          RGWObjVersionTracker& objv_tracker,
+          optional_yield y, const DoutPrefixProvider* dpp,
+          RGWMDLogSyncType type, bool from_remote_zone) override
+  {
+    auto robj = static_cast<MetadataObject*>(obj);
+    auto& info = robj->get_topic_info();
+    auto mtime = robj->get_mtime();
+
```

Comment:
```suggestion
    if (!info.dest.push_endpoint.empty() && info.dest.persistent) {
      int r = rgw::notify::add_persistent_topic(info.name, y);
      if (r < 0) {
        ldpp_dout(dpp, 1) << "Failed to create queue for persistent topic: "
                          << info.name << " with error: " << r << dendl;
        return r;
      }
    }
```
without this code, for the persistent topics the persistent queue is not created on secondary zone.